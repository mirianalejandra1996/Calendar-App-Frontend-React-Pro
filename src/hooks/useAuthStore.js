import { useDispatch, useSelector } from "react-redux";
import calendarApi from "../api/calendarApi";
import { clearErrorMessage, onChecking, onLogin, onLogout, onLogoutCalendar } from "../store";

export const useAuthStore = () => {
  const { status, user, errorMessage } = useSelector((state) => state.auth);
  const dispatch = useDispatch();

  const startLogin = async ({email, password}) => {
        
    console.log({email, password})
        dispatch(onChecking())
        try {
            const { data } = await calendarApi.post('/auth', { email, password })
            localStorage.setItem('token', data.token)
            localStorage.setItem('token-init-date', new Date().getTime()) // In case we would like to do something
            dispatch(onLogin({name: data.name, uid: data.uid}))
        } catch (error) {
            console.log('error', error.message)
            localStorage.removeItem('token')
            localStorage.removeItem('token-init-date')
            dispatch(onLogout('Invalid credentials'))
            setTimeout(() => dispatch(clearErrorMessage()) , 10)
        }
  }

  const startRegister = async ({ name, email, password }) => {
    
    try {
      const { data } = await calendarApi.post('/auth/new', { name, email, password})


      localStorage.setItem('token', data.token)
      localStorage.setItem('token-init-date', new Date().getTime()) // In case we would like to do something
      // with the time
      dispatch(onLogin({name: data.name, uid: data.uid}))

    } catch (error) {
      console.log('error', error)

      // const {error: errorFieds} = error.response.data 
      const { errors, msg } =  error.response.data 

      if (errors) dispatch(onLogout('Please fill all fields correctly'))
      // else if (msg === 'Email not available')  dispatch(onLogout(msg))
      else if (msg)  dispatch(onLogout(msg === 'Email not available' ? 'Email already in use' : msg))

      else dispatch(onLogout(error.message))

      setTimeout(() => dispatch(clearErrorMessage()) , 10)
    }
  }

  const checkAuthToken =  async() => {
    
    const token = localStorage.getItem('token')
    if  (!token) return dispatch(onLogout())

    try {

        const { data } = await calendarApi.get('/auth/renew')

        // * Note: I can create a functionality to set cookies for code optimization
        localStorage.setItem('token', data.token)
        localStorage.setItem('token-init-date', new Date().getTime()) // In case we would like to do something
        // with the time

        dispatch(onLogin({name: data.name, uid: data.uid}))


      } catch (error) {
            console.log('error', error)
            localStorage.clear()
            dispatch(onLogout())
    }
  }

  const startLogout = () => {
    dispatch(onLogout())
    dispatch(onLogoutCalendar())
    localStorage.clear()
  }

  return {
    // * Properties
    status,
    user,
    errorMessage,

    // * Methods
    checkAuthToken,
    startLogin,
    startLogout,
    startRegister,
  };
};
