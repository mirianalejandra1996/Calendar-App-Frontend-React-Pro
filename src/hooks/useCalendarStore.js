import { useDispatch, useSelector } from "react-redux";
import {
  onAddNewEvent,
  onLogoutCalendar,
  onDeleteEvent,
  onLoadEvents,
  onSetActiveEvent,
  onUpdateEvent,
} from "../store/calendar/calendarSlice";
import calendarApi from "../api/calendarApi";
import { convertEventsToDateEvents } from "../helpers";
import Swal from "sweetalert2";

export const useCalendarStore = () => {
  const dispatch = useDispatch();

  const { events, activeEvent } = useSelector((state) => state.calendar);
  const { user } = useSelector((state) => state.auth);

  const setActiveEvent = (calendarEvent) => {
    dispatch(onSetActiveEvent(calendarEvent));
  };

  // No vamos a usar un thunks asíncronos, de frente vamos a disparar acciones síncronas
  const startSavingEvent = async (calendarEvent) => {

    try {
      if (calendarEvent.id) {
        // If user has an ID is because the event needs to be updated
        await calendarApi.put(`/events/${calendarEvent.id}`, calendarEvent);
        dispatch(onUpdateEvent({ ...calendarEvent }));
        Swal.fire('Update', "It has been updated successfully", 'success');
        return;
      } else {
        // Else is creating a new event and adding an unique ID event
        const { data } = await calendarApi.post("/events", calendarEvent);
        dispatch(onAddNewEvent({ ...calendarEvent, id: data.event.id, user: { ...user, _id: user.uid} }));
        Swal.fire('Creation', "It has been created successfully", 'success');
      
      }
    } catch (error) {
      console.log('error', error)
      Swal.fire('Error while saving event', error.response.data?.msg , 'error' )
    }
  };

  const startDeletingEvent = async () => {

    try {
      await calendarApi.delete(`/events/${ activeEvent.id }` );
      dispatch(onDeleteEvent());
      Swal.fire('Elimination', "It has been removed successfully", 'success');
      return;
      
    } catch (error) {
      console.log('error', error)
      Swal.fire('Error while deleting event', error.response.data?.msg , 'error' )
    }
  };

  const startLoadingEvents = async () => {

    try {
      const { data } = await calendarApi.get("/events");

      console.log('loading events', data)
      const events = convertEventsToDateEvents(data.events);

      dispatch(onLoadEvents(events));
    } catch (error) {
      console.log("Error loading events", error);
    }
  };  

  return {
    // * Propiedades
    events,
    activeEvent,
    hasEventSelected: !!activeEvent,

    // * Métodos
    setActiveEvent,
    startDeletingEvent,
    startSavingEvent,
    startLoadingEvents,
  };
};
