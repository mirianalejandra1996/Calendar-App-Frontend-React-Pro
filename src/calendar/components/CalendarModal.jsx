import { useMemo, useState, useEffect } from "react";
import Swal from "sweetalert2";
import "sweetalert2/dist/sweetalert2.min.css";
import { addHours } from "date-fns";

import DatePicker, { registerLocale } from "react-datepicker";
import "react-datepicker/dist/react-datepicker.css";
import es from "date-fns/locale/es";

import Modal from "react-modal";
import { differenceInSeconds } from "date-fns/esm";

import { useAuthStore, useCalendarStore, useUiStore } from "../../hooks";
import { getEnvVariables } from "../../helpers";

registerLocale("es", es);

const initialState = {
  title: "",
  notes: "",
  start: new Date(),
  end: addHours(new Date(), 2),
};

const customStyles = {
  content: {
    top: "50%",
    left: "50%",
    right: "auto",
    bottom: "auto",
    marginRight: "-50%",
    transform: "translate(-50%, -50%)",
  },
};

// Modal.setAppElement("#root");

if (getEnvVariables().VITE_MODE !== 'test') {
  Modal.setAppElement('#root');
}

export const CalendarModal = () => {
  const { isDateModalOpen, closeDateModal } = useUiStore();
  const { activeEvent, startSavingEvent } = useCalendarStore();
  const { user } = useAuthStore();

  const isMyEvent = user?.uid === activeEvent?.user?._id;
  const isNewEvent = !activeEvent?.id;

  const [formSubmitted, setFormSubmitted] = useState(false);

  const [formValues, setFormValues] = useState(initialState);

  useEffect(() => {
    if (activeEvent) {
      setFormValues({ ...activeEvent });
    } else {
      setFormValues(initialState);
    }
  }, [activeEvent]);

  const titleClass = useMemo(() => {
    if (!formSubmitted) return "";

    return formValues.title.length > 0
      ? // ? 'is-valid'
        ""
      : "is-invalid";
  }, [formValues.title, formSubmitted]);

  const onInputChange = ({ target }) => {
    setFormValues({ ...formValues, [target.name]: target.value });
  };

  const onDateChange = (date, name) => {
    setFormValues({ ...formValues, [name]: date });
  };

  const onSubmit = async (e) => {
    e.preventDefault();
    setFormSubmitted(true);
    console.log("onSubmit");

    const difference = differenceInSeconds(formValues.end, formValues.start);

    if (isNaN(difference) || difference <= 0)
      return Swal.fire(
        "Fechas incorrectas",
        "Revisar las fechas ingresadas",
        "error"
      );

    if (formValues.title.length <= 0) return;

    await startSavingEvent(formValues);
    closeDateModal();

    setFormSubmitted(true);
  };

  return (
    <Modal
      isOpen={isDateModalOpen}
      onRequestClose={closeDateModal}
      style={customStyles}
      contentLabel="Example Modal"
      ariaHideApp={false}
      className="modal"
      overlayClassName="modal-fondo"
      closeTimeoutMS={200}
    >
      <h1> Nuevo evento </h1>
      <hr />
      <form className="container" onSubmit={onSubmit}>
        <div className="form-group mb-2">
          <label>Fecha y hora inicio</label>
          <DatePicker
            // name={"start"}
            selected={formValues.start}
            onChange={(e) => onDateChange(e, "start")}
            className="form-control"
            dateFormat="Pp"
            showTimeSelect
            locale="es"
            timeCaption="Hora"
          />
        </div>

        <div className="form-group mb-2">
          <label>Fecha y hora fin</label>
          <DatePicker
            // name={"end"}
            minDate={formValues.start}
            selected={formValues.end}
            onChange={(e) => onDateChange(e, "end")}
            className="form-control"
            dateFormat="Pp"
            showTimeSelect
            locale="es"
            timeCaption="Hora"
          />
        </div>

        <hr />
        <div className="form-group mb-2">
          <label>Titulo y notas</label>
          <input
            type="text"
            className={`form-control ${titleClass}`}
            placeholder="Título del evento"
            name="title"
            autoComplete="off"
            value={formValues.title}
            onChange={onInputChange}
          />
          <small id="emailHelp" className="form-text text-muted">
            Una descripción corta
          </small>
        </div>

        <div className="form-group mb-2">
          <textarea
            type="text"
            className="form-control"
            placeholder="Notas"
            rows="5"
            name="notes"
            value={formValues.notes}
            onChange={onInputChange}
          ></textarea>
          <small id="emailHelp" className="form-text text-muted">
            Información adicional
          </small>
        </div>

        {(isMyEvent || isNewEvent) && (
          <button type="submit" className="btn btn-outline-primary btn-block">
            <i className="far fa-save"></i>
            <span> Guardar</span>
          </button>
        )}
      </form>
    </Modal>
  );
};
