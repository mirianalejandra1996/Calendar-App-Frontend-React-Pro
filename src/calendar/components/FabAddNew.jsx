import { addHours } from "date-fns"
import { useUiStore, useCalendarStore } from "../../hooks"

export const FabAddNew = () => {

    const { openDateModal } = useUiStore()
    const { setActiveEvent } = useCalendarStore()

    const handleClickNew = () => {

        setActiveEvent({
            title: '',
            notes: '',
            start: new Date(),
            end: addHours(new Date(), 1),
            bgColor: '#fafafa',
            // user: {
            //     _id: 123,
            //     name: 'Ale',
            // }
          })
        openDateModal()
      }

  return (
    <button
        onClick={handleClickNew}
        className='btn btn-primary fab'
    >
        <i className='fas fa-plus'></i>
    </button>
  )
}
