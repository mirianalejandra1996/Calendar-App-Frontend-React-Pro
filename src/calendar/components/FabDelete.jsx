import { useAuthStore, useCalendarStore, useUiStore } from "../../hooks"

export const FabDelete = () => {

    const { startDeletingEvent, activeEvent, hasEventSelected } = useCalendarStore()
    const { isDateModalOpen } = useUiStore()
    const { user } = useAuthStore()

    const isMyEvent = user?.uid === activeEvent?.user?._id
    
    const handleDelete = () => {

        if(!activeEvent) return

        startDeletingEvent(activeEvent)
      }

  return (
    <>
    {
      isMyEvent
      ? (
        <button
        aria-label="btn-delete"
        onClick={handleDelete}
        className='btn btn-danger fab-danger'
        style={{display: (hasEventSelected && !isDateModalOpen ) ? '' : 'none'}}
    >
        <i className='fas fa-trash-alt'></i>
    </button>
      )
      : null
    }
    </>
  )
}
