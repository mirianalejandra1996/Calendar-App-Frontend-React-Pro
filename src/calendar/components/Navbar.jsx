import React from 'react'
import { useAuthStore } from '../../hooks'

export const Navbar = () => {

  const {startLogout, user} = useAuthStore()

  const handleLogout = () => {
    startLogout()
  }
  return (
    <div className='navbar navbar-dark bg-dark mb-4 px-4'>
        <span className='navbar-brand capitalize'>
            <i className='fas fa-calendar-alt'></i>
            &nbsp;
            &nbsp;
            {user.name}
        </span>

        <button aria-label='logout-btn' onClick={handleLogout} className='btn btn-outline-danger'>
            <i className='fa fa-sign-out-alt'></i>
            &nbsp;
            <span>Salir</span>
        </button>
    </div>
  )
}
