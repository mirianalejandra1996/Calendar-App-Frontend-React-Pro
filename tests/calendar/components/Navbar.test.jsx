import { fireEvent, render, screen } from "@testing-library/react"
import { Navbar } from "../../../src/calendar/components/Navbar"

import { useAuthStore } from "../../../src/hooks/useAuthStore"
import { demoUser } from "../../fixtures/authFixtures"

jest.mock("../../../src/hooks/useAuthStore")

describe('Tests on <Navbar/>', () => {

    const mockStartLogout = jest.fn()

    beforeEach(() => jest.clearAllMocks())

    test('should display the component correctly if user is logged in', () => {

        useAuthStore.mockReturnValue({
            startLogout: mockStartLogout,
            user: demoUser,
        })

        const { container } = render(<Navbar/>)

        expect(container).toMatchSnapshot()

        // display user name on navbar
        expect(screen.getByText(demoUser.name)).toBeTruthy()

        const logoutBtn = screen.getByLabelText('logout-btn')

        // console.log(logoutBtn)

        fireEvent.click(logoutBtn)

        // screen.debug()

        expect(mockStartLogout).toHaveBeenCalled()
        expect(mockStartLogout).toHaveBeenCalledTimes(1)

    })
})