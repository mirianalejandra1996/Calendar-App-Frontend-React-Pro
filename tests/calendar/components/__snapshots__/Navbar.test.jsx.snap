// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on <Navbar/> should display the component correctly if user is logged in 1`] = `
<div>
  <div
    class="navbar navbar-dark bg-dark mb-4 px-4"
  >
    <span
      class="navbar-brand capitalize"
    >
      <i
        class="fas fa-calendar-alt"
      />
         
      alejandra
    </span>
    <button
      aria-label="logout-btn"
      class="btn btn-outline-danger"
    >
      <i
        class="fa fa-sign-out-alt"
      />
       
      <span>
        Salir
      </span>
    </button>
  </div>
</div>
`;
